src/Export/ProductService.php

```diff
@@ -261,4 +244,50 @@ protected function addProductIdFilters(Criteria $criteria, string $productId): v
             )
         );
     }
+
+    protected function getVisibleProducts(?int $limit, ?int $offset, ?string $productId): EntitySearchResult
+    {
+        $criteria = $this->getCriteriaWithProductVisibility($limit, $offset);
+
+        if ($productId) {
+            $this->addProductIdFilters($criteria, $productId);
+        }
+
+        return $this->container->get('product.repository')->search(
+            $criteria,
+            $this->salesChannelContext->getContext()
+        );
+    }
+
+    protected function buildProductsWithVariantInformation(EntitySearchResult $result): ProductCollection
+    {
+        $products = new ProductCollection();
+
+        /** @var ProductEntity $product */
+        foreach ($result->getEntities() as $product) {
+            if ($product->getMainVariantId() !== null) {
+                $product = $this->getRealMainProductWithVariants($product->getMainVariantId());
+            } else {
+                $this->assignChildrenOrSiblings($product);
+            }
+
+            $products->add($product);
+        }
+
+        return $products;
+    }
+
+    protected function getRealMainProductWithVariants(string $realMainProductId): ProductEntity
+    {
+        $mainProduct = $this->getVisibleProducts(1, 0, $realMainProductId)->first();
+        $this->assignChildrenOrSiblings($mainProduct);
```

Comment:
Still possible to remove when the if above is changed like suggested.

Reply:
Ah this is why you wanted it to be changed. Honestly I have rewritten this whole code about three times to get a good structure, so when I got to this point I didn't see that we could simplify this even further. Nice, thanky.